src/core/stats.h

```diff
@@ -55,6 +59,59 @@ struct StatCounter {
     void log(int count = 1) { Stats::log(id, count); }
 };
 
+struct StatTimer {
+private:
+    static __thread StatTimer* stack;
+
+    Timer _timer;
+    StatTimer* _prev;
+
+    int _statid;
+    bool _paused;
+
+    void logAndPop() {
+        assert(stack == this);
+        if (!_paused)
+            Stats::log(_statid, _timer.end());
+        stack = _prev;
+        if (stack)
+            stack->resume();
+    }
+
+public:
+    StatTimer(int statid) : _timer(NULL), _statid(statid), _paused(false) {
+        if (stack) {
+            stack->pause();
+        }
+        _prev = stack;
+        stack = this;
+    }
+    ~StatTimer() { logAndPop(); }
+
+    void pause() {
+        assert(!_paused);
+        Stats::log(_statid, _timer.end());
+        _paused = true;
+    }
+    void resume() {
+        assert(_paused);
+        _paused = false;
+        _timer.restart(NULL);
+    }
+
+    bool isPaused() const { return _paused; }
+
+    static StatTimer* getStack() { return stack; }
+    static void setStack(StatTimer* s, bool andResume = false) {
+        stack = s;
+        if (andResume && s && s->isPaused()) {
```

Comment:
why wouldn't the other stack be paused?

there are a couple places we check isPaused, but it seems like we should just assert(paused)